### ISSUE TITLE
`task` name is null in the Terraform state when not specified

### ISSUE THREAD

**Issue by DavidGOrtega** [iterative/terraform-provider-iterative#419]
Seems that `name` property is ~useless right now~ is not stored if not specified. It should be stored the generated name to feed the name to keep the state coherence

```json
"name": null,
"parallelism": 1,
```

**Comment by 0x2b3bfa0** [comment]
Name is empty only when you use random identifiers (i.e. leave it blank). If you use deterministic identifiers, it won't be empty.

**Comment by 0x2b3bfa0** [comment]
Where did you get that JSON (?) excerpt from?

**Comment by DavidGOrtega** [comment]
> Name is empty only when you use random identifiers (i.e. leave it blank). If you use deterministic identifiers, it won't be empty.

Whats the purpose of name then? I guess that the machine is retrieved using the Id if no name, If thats the case. Why not just rely in the id?

**Comment by DavidGOrtega** [comment]
> Where did you get that JSON (?) excerpt from?

Running a task without specifying the name

**Comment by 0x2b3bfa0** [comment]
The `name` attribute is the only way of _setting_ a deterministic identifier. Users can't directly specify what `id` a resource will have.

**Comment by DavidGOrtega** [comment]
My biggest concern now is, How do we recognise a task? By `Id` or by `name`?

**Comment by 0x2b3bfa0** [comment]
> _My biggest concern now is, How do we recognise a task? By `Id` or by `name`?_

Tasks are always guaranteed to have an `id` upon creation. The `name` argument is only useful for deterministic runs in CI/CD scenarios.

**Comment by DavidGOrtega** [comment]
I think that we should just generate a `name` and proceed like the name were given by the user. That way everything behaves the same and the outcome is always the same

**Comment by 0x2b3bfa0** [comment]
[Computed](https://www.terraform.io/plugin/sdkv2/schemas/schema-behaviors#computed) values can't be overriden by the user, and user–specified values shouldn't be computed in unstable ways. That's why we have an optional `name` attribute to seed the identifier generator.

**Comment by DavidGOrtega** [comment]
What I say is that if the user do not set a name we should generate one (to seed that computed identifier generator) also and have a name property useful and meaningful

**Comment by 0x2b3bfa0** [comment]
> _What I say is that if the user do not set a name we should generate one (to seed that computed identifier generator also)_

We already do that

**Comment by DavidGOrtega** [comment]
> We already do that, just that it doesn't get saved to the name attribute.

Nice! it should be saved to keep coherence

**Comment by DavidGOrtega** [comment]
☝️ updated issue description

**Comment by casperdcl** [comment]
> We already do that, just that it doesn't get saved to the name attribute.

is there a reason it *shouldn't* be saved, @0x2b3bfa0?

**Comment by 0x2b3bfa0** [comment]
When a provider sets a non–computed attribute in the state to values other than the one specified in `*.tf` files, any later plan/apply operations will trigger a `"value" ~> null` diff and recreate the resource even when it's unchanged.

TL;DR if you're querying the state, use always `id`

**Comment by casperdcl** [comment]
Ok so this issue should be closed by a documentation note?

**Comment by 0x2b3bfa0** [comment]
I believe so

**Comment by DavidGOrtega** [comment]
> any later plan/apply operations will trigger a "value" ~> null diff and recreate the resource even when it's unchanged.

I think that we can update it during the creation without triggering the recreation. I will give it a try

**Comment by 0x2b3bfa0** [comment]
Not possible unless proven otherwise
